skimage/exposure/_adapthist.py

```diff
@@ -122,74 +123,103 @@ def _clahe(image, kernel_size, clip_limit, nbins):
     minimum and maximum value as the input image. A clip limit smaller than 1
     results in standard (non-contrast limited) AHE.
     """
-    row_step = int(image.shape[0] // np.ceil(image.shape[0] / kernel_size[0]))
-    col_step = int(image.shape[1] // np.ceil(image.shape[1] / kernel_size[1]))
+    # pad the image so that the shape in each dimension
+    # is a multiple of the relevant kernel_size
+    pad_end_per_dim = [0] * image.ndim
+    for dim in range(image.ndim):
+        pad_end_per_dim[dim] = (kernel_size[dim] - image.shape[dim] % kernel_size[dim]) % kernel_size[dim]
 
-    nr = image.shape[0] // row_step
-    nc = image.shape[1] // col_step
+    image = np.pad(image,
+                   [[0, pad_end_per_dim[dim]] for dim in range(image.ndim)],
+                   mode='reflect',
+                   )
+
+    ns = [int(np.ceil(image.shape[dim] / kernel_size[dim]))
+          for dim in range(image.ndim)]
+
+    steps = [int(np.floor(image.shape[dim] / ns[dim]))
+             for dim in range(image.ndim)]
 
     bin_size = 1 + NR_OF_GRAY // nbins
     lut = np.arange(NR_OF_GRAY)
     lut //= bin_size
 
-    map_array = np.zeros((nr, nc, nbins), dtype=int)
+    map_array = np.zeros(tuple(ns) + (nbins,), dtype=int)
 
     # Calculate graylevel mappings for each contextual region
-    i0 = 0
-    for r in range(nr):
-        i1 = i0 + row_step
-        j0 = 0
-        for c in range(nc):
-            j1 = j0 + col_step
-            sub_img = image[i0:i1, j0:j1]
-
-            if clip_limit > 0.0:  # Calculate actual cliplimit
-                clim = max(int(clip_limit * sub_img.size), 1)
+    for inds in np.ndindex(*ns):
+
+        region = tuple([slice(inds[dim] * steps[dim], (inds[dim] + 1)
+                              * steps[dim])
+                        for dim in range(image.ndim)])
+        sub_img = image[region]
+
+        if clip_limit > 0.0:  # Calculate actual clip limit
+            clim = int(clip_limit * sub_img.size)
+            if clim < 1:
+                clim = 1
+        else:
+            clim = NR_OF_GRAY  # Large value, do not clip (AHE)
+
+        hist = lut[sub_img.ravel()]
+        hist = np.bincount(hist)
```

Comment:
why did you change the existing code `np.bincount(hist, minlength=nbins)` ? It looked ok to me.

Reply:
You're right, the existing line here is better and should stay! It's a change from three months ago (https://github.com/scikit-image/scikit-image/commit/834f8b6701ec55394f0e48d59a1524a4bbada14a) which happened after I had copied the code to adapt it. This was actually some years ago already and recently I went through new commits but seem to have missed that one. Just changed the line back to master. Thanks!